app/javascript/src/components/_token.scss

```diff
@@ -1,23 +1,30 @@
+:root {
+  --zoom-scale: 1;
+}
+
 .token {
   @apply
     flex
     items-center
     absolute;
-  width: min-content;
+
+  margin-left: calc(-1rem * var(--zoom-scale));
+  margin-top: calc(-1rem * var(--zoom-scale));
 }
 
 .token__image {
   @apply
-    w-8
-    h-8
     bg-cover
     bg-center
     rounded-full
-    border-2
     border-white
     pointer-events-none
     z-0
     shadow;
+
+  border-width: calc(2px * var(--zoom-scale));
```

Comment:
Perhaps a thought for another day, but this makes me wonder if there are other things we could do in pure css that we're currently doing in javascript.

Reply:
Yeah I wanted to avoid doing too much layout and styling in JS. I like Stimulus' ethos of using markup to attach behaviour.

One cool thing I found a while ago was that you can set variables in style tags so you can make this sort of thing work:

```html
<style>
.token {
  left: var(--x);
  right: var(--y);
}
</style>

<div class="token" style="--x: 120; --y: 350;"></div>
```

Reply:
I've been thinking even more about this and wondered about adding functionality to mirror data attributes to css variables?

That way, this will work automatically:

```html
<style>
.token {
  left: var(--data-x)px;
  top: var(--data-y)px;
}
</style>

<div class="token" data-x="120" data-y="350"></div>
```

Hopefully it would mean we could move all style definitions to CSS and let the JS just manipulate the content of the DOM.

Reply:
Stimulus uses [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to manage this kind of thing so we could look at adding this to controllers quite easily.

Reply:
Something like this in the controller:

```js
connect() {
  this.observer = new MutationObserver(mutations => {
    mutations.forEach(mutation => {
      if (mutation.attributeName.startsWith("data-")) {
        mutation.target.setStyle(`--${mutation.attributeName}`, target.dataset[mutation.attributeName])
      }
    })
  })

  this.observer.observe(this.targetNode, { attributes: true })
}

disconnect() {
  this.observer.disconnect()
}
```

Could either be initialised for individual targets, or could be initialised for all targets defined in `static targets = []`